lib/utils/dotTree.ts

```diff
@@ -0,0 +1,177 @@
+label=<LatchedSelector<BR/><B>${LatchedSelectorSymbol}</B>>];
+	LatchedSequence [${CompositeShape} ${NodeStyle} ${DefaultColor} \
+label=<LatchedSequence<BR/><B>${LatchedSequenceSymbol}</B>>];
+	IfElse [${CompositeShape} ${NodeStyle} ${DefaultColor} label=<IfElse<BR/><B>${IfElseSymbol}</B>>];
+	Not [${DecoratorShape} ${NodeStyle} ${DefaultColor} label=<Not<BR/><B>${NotSymbol}</B>>];
+	RepeatWhen [${DecoratorShape} ${NodeStyle} ${DefaultColor} label=<RepeatWhen<BR/><B>${RepeatWhenSymbol}</B>>];
+	RepeatOnResult [${DecoratorShape} ${NodeStyle} ${DefaultColor} \
+label=<RepeatOnResult<BR/><B>${RepeatOnResultSymbol}</B>>];
+	ResultSwapSymbol [${DecoratorShape} ${NodeStyle} ${DefaultColor} \
+label=<ResultSwapSymbol<BR/><B>${ResultSwapSymbol}</B>>];`;
+
+export function serializeColorLegend(): string {
+	return `digraph ColorLegend {
+${nodeColorLegend}
+}`;
+}
+
+export function serializeShapeLegend(): string {
+	return `digraph ShapeLegend {
+${nodeShapeLegend}
+}`;
+}
+
+export function serializeSymbolLegend(): string {
+	return `digraph SymbolLegend {
+${nodeSymbolLegend}
+}`;
+}
+
+function getShape<S extends BlueshellState, E>(node: Base<S, E>): string {
+	if (node instanceof Decorator) {
+		return DecoratorShape;
+	} else if (node instanceof Composite) {
+		return CompositeShape;
+	} else {
+		return DefaultShape;
+	}
+}
+
+function getColor<S extends BlueshellState, E>(node: Base<S, E>, state?: S): string {
+	if (state) {
+		const eventCounter = node!.getTreeEventCounter(state);
+		const lastEventSeen = node!.getLastEventSeen(state);
+		const lastResult = node!.getLastResult(state);
+
+		if (lastEventSeen === eventCounter && lastResult) {
+			switch (lastResult) {
+			case rc.ERROR:
+				return ErrorColor;
+			case rc.SUCCESS:
+				return SuccessColor;
+			case rc.RUNNING:
+				return RunningColor;
+			case rc.FAILURE:
+				return FailureColor;
+			}
+		}
+	}
+	return DefaultColor;
+}
+
+function getNodeId<S extends BlueshellState, E>(node: Base<S, E>): string {
+	let id = node.name;
+	if (id !== node.constructor.name) {
+		id += '_' + node.constructor.name + '_';
+	}
+	return id;
+}
+
+function getLabel<S extends BlueshellState, E>(node: Base<S, E>): string {
+	const displayText = node.name;
+	let nodeSymbol: string;
+	if (node instanceof LatchedSelector) {
+		nodeSymbol = LatchedSelectorSymbol;
+	} else if (node instanceof Selector) {
+		nodeSymbol = SelectorSymbol;
+	} else if (node instanceof LatchedSequence) {
+		nodeSymbol = LatchedSequenceSymbol;
+	} else if (node instanceof Sequence) {
+		nodeSymbol = SequenceSymbol;
+	} else if (node instanceof IfElse) {
+		nodeSymbol = IfElseSymbol;
+	} else if (node instanceof Not) {
+		nodeSymbol = NotSymbol;
+	} else if (node instanceof RepeatOnResult) {
+		nodeSymbol = RepeatOnResultSymbol;
+	} else if (node instanceof RepeatWhen) {
+		nodeSymbol = RepeatWhenSymbol;
+	} else if (node instanceof ResultSwap) {
+		nodeSymbol = ResultSwapSymbol;
```

Comment:
could / should this be supplied by the node itself? since its unicode it could be used by the archy serializer as well:

`node.symbol`

Reply:
I was originally opposed, but I like it now for any inherited class can set it's own symbol.